Implement microtonality (per-note tuning)

jean-emmanuel opened the pull request:
This might be a naive way to do so, but it seems to work fine so far. Now we can sing in tune a detuned synth ;). Feel free to discard I you prefer not to add new features.

Edit: I have not added the fields in the ui

jean-emmanuel committed:
Implement microtonality (per-note tuning)

x42 commented:
Looks like you've only patched the JACK app. I expect it will crash if you load the LV2 plugin.

From a compatibility point of view it would be a lot nicer to re-use the existing control-ports (`m00` .. `m11`). They are currently  integers 0 (off), 1(enable)  but could be a float range:  0..2. (difference from 1.0 would be the detune amount). 

```
if (*self->port [FAT_NOTE + i] > 0) {
        self->retuner->set_notescale (i, *self->port [FAT_NOTE + i] -1.f);
}
```

Also note that  `lv2ttl/fat1.h` is auto-generated from the the LV2 plugin description, `lv2ttl/fat1.ttl.in`. remving the `lv2:portProperty lv2:integer, lv2:toggled;` and allowing a range 0,2 should do the trick.

That way you do not have to add any new control-ports and break existing sessions.

As usual, the hard part will be the GUI. I could envisage a small slider on each of the keyboard-keys to set  a detune amount, which is also sensitive in MIDI-mode. Any thoughts?

Then again it could also remain a hidden feature (generic plugin GUI, or jack/OSC only). It's not a very common thing to detune individual notes.

jean-emmanuel commented:
Thank you for reviewing the PR !

Reusing `m00..m11` this way sounds smart but wouldn't the generic GUI become less clear then ?

I'm at loss to find out how to generate `fat1.h`: `ttl2c lv2ttl/fat1.ttl.in  lv2ttl/fat1.h` throws ` Could not parse input. Not valid Turtle syntax.`. Did I miss something ?

As for the openGL GUI I think sliders on keyboard keys might be a bit confusing, it's not possible to put them in a modal window, is it ? Generic GUI / OSC only seems reasonable to me.

x42 commented:
> Reusing m00..m11 this way sounds smart but wouldn't the generic GUI become less clear then ?

hmm, perhaps, yes. To hit 0, 1 exactly scale-points can be used (most hosts use detents on sliders at scale-points).

> I'm at loss to find out how to generate fat1.h: ttl2c lv2ttl/fat1.ttl.in lv2ttl/fat1.h throws Could not parse input. Not valid Turtle syntax.. Did I miss something ?

It takes the plugin's URI (not the file). you'll have to deploy the plugin first (I usually sym-link `~/.lv2/fat1.lv2 -> ....src-dir/build`). Effectively you can just change it manually. I just wanted to mention that the actual source is the .ttl and that the descriptions in the files are supposed to be in sync.

> As for the openGL GUI I think sliders on keyboard keys might be a bit confusing, it's not possible to put them in a modal window, is it ? 

While possible it should generally be avoided at all cost for plugin-UIs: It is really bad design and discouraged by all host authors. --  An expander or overlay (same window) can work.

> Generic GUI / OSC only seems reasonable to me.

There may still be some small GUI changes needed. the GUI could remember custom values. If you switch between generic and custom UI, the custom UI may overwrite the settings with hard 0,1.

x42 commented:
PS. it is possible to add ports to a LV2 plugin -- but only by appending them at the end but that does not strike me as a good solution.

Another good alternative would be to _fork_ the plugin: Have it come in two different variants built from the same source: Like many plugins come in mono and stereo variants.

The _extended_ version would have a different URI  - perhaps a suffix `#microtonal` and could have dedicated ports (like your current prototype).

x42 commented on src/retuner.cc:
```diff
@@ -34,7 +34,8 @@ Retuner::Retuner (int fsamp) :
     _corroffs (0.0f),
-    _notemask (0xFFF)
+    _notemask (0xFFF),
+    _notescale {0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f, 7.0f, 8.0f, 9.0f, 10.0f, 11.0f}
```
This extended initializer list is only available with C++11 or later.
The plugin works on much older systems.  Check with `make CXXFLAGS="-std=c++98"`

jean-emmanuel committed:
Fix compilation for c++98 (don't use c++11 list initializer)

jean-emmanuel committed:
Add microtonal ports to ttl

jean-emmanuel commented:
Thanks for all the explanations :). I've fixed the c++98 compatibility issue and added the ports to the lv2 plugin which loads fine now (in Ardour). The fork way sounds good to me: to sum up I'd need to
- wrap the lines I've added in conditionnal blocks (`#ifdef MICROTONAL`)
- create an alternate ttl file with the extra ports
- change the the makefile so that it use thes alternate ttl if the env variable is found (eg `MICROTONAL`)

x42 commented:
What you propose would be an either/or solution (compile time define).

I was thinking of a runtime decision: the same plugin.so provides both (like mono/stereo versions of other plugin -- here: provides the old and a microtonal variant in the same .so)

jean-emmanuel committed:
Create an additionnal lv2 descriptor to expose both vanilla and microtonal versions of the plugin

jean-emmanuel committed:
Allow default lv2 UI with #microtonal URI too

jean-emmanuel committed:
Make microtonal true by default to enable it in standalone mode

jean-emmanuel commented:
Alright ! Looks good so far, there're just the overlay knobs/sliders missing in the open gl ui (although it doesn't seem to mess with the generic ui).

jean-emmanuel committed:
Make microtonal true by default to enable it in standalone mode (missing linein previous commit)

jean-emmanuel committed:
Fix checking for standalone instance to enable microtonal

x42 commented:
LV2 changes look good. It's a lot nicer to use a separate plugin URI for this!

jean-emmanuel commented:
I haven't had the time to dive into the LV2 gui code yet, I'll try someday soon (hopefully!)

x42 commented:
Are you using the jack-app with OSC remote control? Otherwise it doesn't currently make sense to expose the micro-tonal ports there.

jean-emmanuel commented:
I use the jack app with osc remote control and the plugin version but the generic ui exposes the micro tonal settings there already.

x42 commented:
I've rebased the commits and consolidated some code, in particular the `.ttl`
For consistency I've updated the jack-application to also expose chromatic version (default is microtonal).

jean-emmanuel commented:
Thank you very much !
